middleman_protocol/middleman_protocol/helpers.py

```diff
@@ -0,0 +1,101 @@
+import socket
+
+from typing import Iterator
+from typing import Optional
+
+from middleman_protocol.constants import ESCAPE_CHARACTER
+from middleman_protocol.constants import ESCAPE_SEQUENCES
+from middleman_protocol.constants import FRAME_SEPARATOR
+from middleman_protocol.constants import RECEIVE_BYTES_PER_LOOP
+from middleman_protocol.message import AbstractMiddlemanMessage
+from middleman_protocol.utils import append_frame_separator
+from middleman_protocol.utils import escape_decode_raw_message
+from middleman_protocol.utils import escape_encode_raw_message
+
+
+def split_stream(connection: socket.socket) -> Iterator[bytes]:
+    """
+    Lowest level receiver which determines if received data is frame separator,
+    which means previously received data should be yielded,
+    or if received data is something else and receiver should continue to gather bytes.
+    """
+    assert isinstance(connection, socket.socket)
+
+    received_data = b''
+
+    while True:
+        next_bytes = connection.recv(RECEIVE_BYTES_PER_LOOP)
+
+        for next_byte in next_bytes:
+            if next_bytes == FRAME_SEPARATOR:
+                yield received_data
+                received_data = b''
+            else:
+                received_data += bytes([next_byte])
+
+
+def unescape_stream(connection: socket.socket) -> Iterator[Optional[bytes]]:
+    """
+    Middle level receiver which determines if received data has broken escaping and yields None to indicate it,
+    or reverses escaping and yields unescaped frame.
+    """
+    assert isinstance(connection, socket.socket)
+
+    for unescaped_data in split_stream(connection=connection):
+        assert isinstance(unescaped_data, bytes)
+
+        # If there is an escape character which is not a valid escape sequence,
+        # yield special value indicating a frame with broken escaping and continue the outer loop.
+        # This can be checked by comparing sum of occurrences of escape character
+        # to sum of occurrences of escape sequences. If they are equal, data is escaped correctly.
+        if (
+            unescaped_data.count(ESCAPE_CHARACTER) !=
+            sum([unescaped_data.count(escape_sequence) for escape_sequence in ESCAPE_SEQUENCES.values()])
+        ):
+            yield None
```

Comment:
I thought `None` would be good value, that's why I have chosen it ;) It is also documented in docstring so anyone expectations should be verified there. But seriously, I think `None` is better than string, even if it would be defined as a constant, because it then would have to be reused (and imported) all over the place where helper is used, while `None` is built in, already available.

Btw. should it be just `yield` without anything after it ?

Reply:
Sorry, I meant `('malformed frame', )` (i.e. a tuple that contains a string).

I still think that `None` might not be obvious enough. But, well, the alternative is not all that appealing either...

@dybi What do you think?

> Btw. should it be just yield without anything after it ?

Definitely `None`. It's the same rule as with `return`.